### DISCUSSION TITLE
Vectorization and Features

### DISCUSSION THREAD

**Discussion by Doubl3MD** [halide/Halide#7643] [Q&A] [answered]
Hi Hiladians,

Forgive my rather n00bish question here, but I want to learn...

I've noticed the following.

When creating a Target from scratch with **_no features_**, for example 'Win-x86-64bit', what does the vectorize() method do?
I've tested the following standard schedule(should look familiar.. :) : 

```
Var x_outer, y_outer, x_inner, y_inner, tile_index;
   _output
      .tile(x, y, x_outer, y_outer, x_inner, y_inner, 64, 64)
      .fuse(x_outer, y_outer, tile_index)
      .parallel(tile_index);
Var x_inner_outer, y_inner_outer, x_vectors, y_pairs;
   _output
      .tile(x_inner, y_inner, x_inner_outer, y_inner_outer, x_vectors, y_pairs, 8, 8)
      .vectorize(x_vectors) // <== commenting and uncommenting this line
      .unroll(y_pairs);
```
No Features for vectorizing where enabled. Yet it goes up to +15% faster if the `.vectorize(x_vectors)` is active.
I would expect no vectorization because no features are enabled... 

What is going on here?
What causes the speedgain if no vectorization feature is set?
Is there a standard vectorizing method? 
Does it influence the unroll that is scheduled next?

Thx in advance for some clarifications.

Regards,

Doubl3MD

**Comment by Doubl3MD** [comment]
Update: While testing other pipes. The speed gains are sometimes up to 50% and more... probably depending on what the algo is doing.
Even more now I'd like to know why with and without the vectorize() line is doing if no vectorize simd features is enabled.

**Comment by steven-johnson** [comment, marked as answer]
Halide assumes that all x86 targets have (at least) SSE2 as a baseline; that provides some modest but nonzero vectorization opportunity. Could that be what you are seeing?

**Comment by Doubl3MD** [comment]
That is indeed what I'm seeing. The algo's I tested showed improvements in line with what sse2 offers.
